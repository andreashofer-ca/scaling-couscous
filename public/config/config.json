{
  "baseUrl": "http://localhost:3000/local-proxy",
  "oidc": {
    "clientId": "fa63bd68-cf20-48e1-a0c1-61716f9aae06",
    "authority": "https://login.microsoftonline.com/f9aeb8ae-3ad2-4947-a6c0-f606f32f91bb/v2.0",
    "scope": "https://appfioneer.onmicrosoft.com/app-ep-sso/service_api openid profile offline_access",
    "store": "local"
  }
}
