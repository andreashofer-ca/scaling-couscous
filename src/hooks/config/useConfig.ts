import { createContext } from 'react'
import { getConfig } from 'api/Config'
import { useQuery } from '@tanstack/react-query'

/**
 * Hook API for accessing the configuration
 * @returns {{isPending: boolean, isError: boolean, data: <config object>}}
 */
export const useConfig = () => {
  const destination =
    import.meta.env.NODE_ENV === 'production' ? '/config/config.json' : '/config/local.config.json'

  const { data, isPending, isError } = useQuery<any>({
    queryKey: ['config'],
    queryFn: async () => getConfig(destination),
    staleTime: Infinity,
    refetchInterval: false,
  })

  return { data, isPending, isError }
}

export const ConfigContext = createContext({})
