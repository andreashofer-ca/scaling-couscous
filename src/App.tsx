import { Home } from 'pages/Home'
import { AuthProvider } from 'react-oidc-context'
import { ConfigContext, useConfig } from 'hooks/config/useConfig'
import buildOidcConfiguration from 'components/ui/authentication/oidcConfigBuilder'
import AuthenticationWrapper from 'components/ui/authentication/AuthenticationWrapper'
import PendingScreen from 'components/ui/screens/PendingScreen'
import { useTranslation } from 'react-i18next'
import 'i18n/setupi18n'

function App() {
  const { t } = useTranslation()

  const { data: config, isPending, isError } = useConfig()

  if (isPending) {
    return (
      <PendingScreen
        title={t('app.loading.in-progress')}
        description={t('app.loading.in-progress')}
      />
    )
  }

  if (isError || !config) {
    return <div>{t('app.loading.error')}</div>
  }

  const { oidc: oidcConfig } = config

  return (
    <ConfigContext.Provider value={config}>
      <AuthProvider {...buildOidcConfiguration(oidcConfig)}>
        <AuthenticationWrapper>
          <Home />
        </AuthenticationWrapper>
      </AuthProvider>
    </ConfigContext.Provider>
  )
}

export default App
