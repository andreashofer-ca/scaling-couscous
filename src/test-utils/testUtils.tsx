import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import { render, RenderOptions } from '@testing-library/react'
import { ThemeProvider } from '@ui5/webcomponents-react'
import React, { ReactElement } from 'react'
import { BrowserRouter } from 'react-router-dom'
import { ConfigContext } from 'hooks/config/useConfig'
import config from 'test/config'

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  })
  return (
    <QueryClientProvider client={queryClient}>
      <ThemeProvider>
        <ConfigContext.Provider value={config}>
          <BrowserRouter>{children}</BrowserRouter>
        </ConfigContext.Provider>
      </ThemeProvider>
    </QueryClientProvider>
  )
}

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, 'wrapper'>) =>
  render(ui, { wrapper: AllTheProviders, ...options })

const AllTheProvidersWithoutBrowserRouter = ({ children }: { children: React.ReactNode }) => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  })
  return (
    <QueryClientProvider client={queryClient}>
      <ThemeProvider>
        <ConfigContext.Provider value={config}>{children}</ConfigContext.Provider>
      </ThemeProvider>
    </QueryClientProvider>
  )
}

const customRenderWithoutBrowserRouter = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>,
) => render(ui, { wrapper: AllTheProvidersWithoutBrowserRouter, ...options })

export * from '@testing-library/react'
export { customRender as render }

export { customRenderWithoutBrowserRouter as renderWithoutRouter }
