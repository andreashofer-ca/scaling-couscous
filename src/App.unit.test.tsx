import { Mock, vi, describe, afterEach, test, expect } from 'vitest'
import App from 'App'
import { useConfig } from 'hooks/config/useConfig'
import { render, screen } from 'test-utils/testUtils'

vi.mock('react-oidc-context')

vi.mock('hooks/config/useConfig')
const mockedUseConfig = useConfig as Mock

describe('App', () => {
  afterEach(() => {
    vi.clearAllMocks()
  })

  test('renders pending message when config is still pending', () => {
    mockedUseConfig.mockReturnValue({
      isPending: true,
      isError: false,
      data: {},
    })

    render(<App />)

    expect(screen).toMatchSnapshot()
  })

  test('renders error message when config has failed to load', () => {
    mockedUseConfig.mockReturnValue({
      isPending: false,
      isError: true,
      data: null,
    })

    render(<App />)

    expect(screen).toMatchSnapshot()
  })

  test('renders app content when config has loaded', () => {
    mockedUseConfig.mockReturnValue({
      isPending: false,
      isError: false,
      data: {
        baseUrl: 'http://localhost:8080/server_info',
        oidc: {
          clientId: '233',
          authority: 'https://authority-test',
          scope: 'openid profile offline_access email user.read',
          store: 'local',
        },
      },
    })

    render(<App />)

    expect(screen).toMatchSnapshot()
  })
})
