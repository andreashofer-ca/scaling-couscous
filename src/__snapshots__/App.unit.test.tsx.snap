// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > renders app content when config has loaded 1`] = `
{
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "logTestingPlaygroundURL": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
}
`;

exports[`App > renders error message when config has failed to load 1`] = `
{
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "logTestingPlaygroundURL": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
}
`;

exports[`App > renders pending message when config is still pending 1`] = `
{
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "logTestingPlaygroundURL": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
}
`;
